spring-session/src/main/java/org/springframework/session/IndexResolver.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2014-2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.session;
+
+/**
+ * Strategy interface for resolving the value of the {@link Session}'s index.
+ *
+ * @param <S> the type of Session being handled
+ * @author Vedran Pavic
+ * @since 1.3.0
+ * @see FindByIndexNameSessionRepository
+ */
+public interface IndexResolver<S extends Session> {
+
+	/**
+	 * Resolve the value of the given index from the session.
+	 * @param session the session
+	 * @param indexName the index name to resolve
+	 * @return the resolved index value, {@code null} if none found
+	 */
+	String resolve(S session, String indexName);
```

Comment:
Each ```FindByIndexNameSessionRepository``` implementation would internally define its supported ```IndexResolver```s, either using a ```CompositeIndexResolver``` or some other way. These would then get consumed according to repository needs, i.e. the same way duplicated principal name resolving logic is consumed now.

This approach allows having some resolvers that are reused across all repository implementations, such as ```PrincipalNameIndexResolver```, as well as each implementation having its specific resolvers. If you would go the _input should just be the session and the result that has a mapping of index name to index value_ route then you wouldn't be able to reuse this logic.